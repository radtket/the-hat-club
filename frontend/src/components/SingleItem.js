import React from "react";
import { Link } from "react-router-dom";
import { margin, cover, size, position } from "polished";
import styled from "styled-components";
import PropTypes from "prop-types";
import { formatMoney } from "../utils/helpers";

const Styles = styled.figure`
  ${margin("48px", null)}

  &:hover {
    /* img {
      transform: translate3d(0, -6px, 0);
    } */

    figcaption {
      dt {
        transform: translateY(-100%);
      }

      dd {
        top: 0;

        > a {
          &::after {
            /* width: 100%; */
            opacity: 1;
            transform: scaleX(1);
          }
        }
      }
    }
  }

  > a {
    display: block;
    position: relative;

    &:hover {
      img {
        &:last-of-type {
          opacity: 1 !important;
        }
      }
    }

    img {
      backface-visibility: hidden;
      display: block;
      margin: 0;
      transition: 0.3s ease;

      &:last-of-type {
        ${cover()}
        opacity: 0;
        transition: 0.5s;
      }
    }
  }

  figcaption {
    color: #242424;
    margin: auto;
    max-width: 315px;
    text-align: center;

    h3,
    dl {
      font-size: 18px;
    }

    h3 {
      ${margin(0, null, "6px")}
      display: -webkit-box;
      -webkit-line-clamp: 2;
      -webkit-box-orient: vertical;
      overflow: hidden;
    }

    dl {
      margin: 0 auto;
      overflow: hidden;
      position: relative;
    }

    dt {
      font-family: "Maisonneue Bold";
      transition: transform 0.3s cubic-bezier(0.25, 0.46, 0.45, 0.94);
    }

    dd {
      ${position("absolute", "100%", null, null, 0)}
      font-size: 16px;
      line-height: 1.7;
      margin-left: 0;
      transition: all 0.3s cubic-bezier(0.25, 0.46, 0.45, 0.94);
      width: 100%;

      > a {
        position: relative;
        color: #747474;

        &::after {
          ${position("absolute", null, null, "-2px", 0)}
          ${size("1px", "100%")}
          content: "";
          background: #747474;
          opacity: 0;
          transform: scaleX(0);
          /* transition: width 0.3s cubic-bezier(0.25, 0.46, 0.45, 0.94) 0.33s; */
          transition: transform 0.3s cubic-bezier(0.25, 0.46, 0.45, 0.94) 0.33s,
            opacity 0.3s cubic-bezier(0.25, 0.46, 0.45, 0.94) 0.33s;
        }
      }
    }
  }
`;

const SingleItem = ({ id, title, price, images }) => {
  const [primaryImage, secondaryImage] = images;
  return (
    <Styles>
      <Link to={`/item/${id}`}>
        <img alt={title} src={primaryImage.largeImage} />
        <img alt={title} src={secondaryImage.largeImage} />
      </Link>
      <figcaption>
        <h3>{title}</h3>
        <dl>
          <dt>{formatMoney(price)}</dt>
          <dd>
            <Link to={`/item/${id}`}>View Details</Link>
          </dd>
        </dl>
      </figcaption>
    </Styles>
  );
};

SingleItem.propTypes = {
  id: PropTypes.string.isRequired,
  title: PropTypes.string.isRequired,
  price: PropTypes.number.isRequired,
  images: PropTypes.arrayOf(
    PropTypes.shape({
      image: PropTypes.string.isRequired,
      largeImage: PropTypes.string.isRequired,
    })
  ).isRequired,
};

export default SingleItem;
