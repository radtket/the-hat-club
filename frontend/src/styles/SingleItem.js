import styled from "styled-components";
import { margin, cover, size, position } from "polished";
import { white, gray, red } from "../utils/colors";
import { rgba } from "../utils/helpers";
import { MoveUpDown } from "./Animations";

export const StyledSingleItemSlideOutNav = styled.div`
  ${margin("48px", null)}
  position: relative;
  overflow: hidden;

  &:hover {
    > a {
      border-bottom-right-radius: 24px;
      border-bottom-left-radius: 24px;
      box-shadow: 0px 15px 10px -15px ${rgba(gray[400], 0.45)};
    }

    img {
      /* transform: translate3d(0, -6px, 0); */
      animation: ${MoveUpDown} 1.5s linear infinite;
    }

    figcaption {
      transform: translateY(-100%);
    }

    nav {
      bottom: -45px;
      box-shadow: 0px 15px 10px -15px ${gray[900]};
    }
  }

  img {
    backface-visibility: hidden;
    display: block;
    transition: 0.3s ease;
  }

  > a {
    background: ${white};
    display: block;
    overflow: hidden;
    position: relative;
    transition: border-radius 0.3s;
    z-index: 2;
  }

  figcaption {
    display: block;
    position: relative;
    transform: translateY(0);
    transition: transform 0.3s;
  }

  dl {
    color: ${gray[800]};
    font-size: 18px;
    margin: auto;
    max-width: 315px;
    text-align: center;

    dt {
      margin-bottom: 6px;
      display: -webkit-box;
      -webkit-line-clamp: 2;
      -webkit-box-orient: vertical;
      overflow: hidden;
    }

    dd {
      font-family: "Maisonneue Bold";
      margin: 0;
    }
  }

  nav {
    background: ${gray[800]};
    border-bottom-left-radius: 24px;
    border-bottom-right-radius: 24px;
    bottom: 100%;
    display: flex;
    justify-content: space-around;
    padding: 36px 12px 12px;
    position: absolute;
    transition: bottom 0.3s;
    width: 100%;

    button {
      &:hover {
        svg {
          fill: ${red[400]};
        }
      }
    }

    svg {
      display: block;
      fill: ${white};
      height: 18px;
    }
  }
`;

export const StyledSingleItem = styled.figure`
  ${margin("48px", null)}

  &:hover {
    /* img {
      transform: translate3d(0, -6px, 0);
    } */

    figcaption {
      dt {
        transform: translateY(-100%);
      }

      dd {
        top: 0;

        > a {
          &::after {
            /* width: 100%; */
            opacity: 1;
            transform: scaleX(1);
          }
        }
      }
    }
  }

  > a {
    display: block;
    position: relative;

    &:hover {
      img {
        &:last-of-type {
          opacity: 1 !important;
        }
      }
    }

    img {
      backface-visibility: hidden;
      display: block;
      margin: 0;
      transition: 0.3s ease;

      &:last-of-type {
        ${cover()}
        opacity: 0;
        transition: 0.5s;
      }
    }
  }

  figcaption {
    color: ${gray[800]};
    margin: auto;
    max-width: 315px;
    text-align: center;

    h3,
    dl {
      font-size: 18px;
    }

    h3 {
      ${margin(0, null, "6px")}
      display: -webkit-box;
      -webkit-line-clamp: 2;
      -webkit-box-orient: vertical;
      overflow: hidden;
    }

    dl {
      margin: 0 auto;
      overflow: hidden;
      position: relative;
    }

    dt {
      font-family: "Maisonneue Bold";
      transition: transform 0.3s cubic-bezier(0.25, 0.46, 0.45, 0.94);
    }

    dd {
      ${position("absolute", "100%", null, null, 0)}
      font-size: 16px;
      line-height: 1.7;
      margin-left: 0;
      transition: all 0.3s cubic-bezier(0.25, 0.46, 0.45, 0.94);
      width: 100%;

      > a {
        position: relative;
        color: ${gray[500]};

        &::after {
          ${position("absolute", null, null, "-2px", 0)}
          ${size("1px", "100%")}
          content: "";
          background: ${gray[500]};
          opacity: 0;
          transform: scaleX(0);
          /* transition: width 0.3s cubic-bezier(0.25, 0.46, 0.45, 0.94) 0.33s; */
          transition: transform 0.3s cubic-bezier(0.25, 0.46, 0.45, 0.94) 0.33s,
            opacity 0.3s cubic-bezier(0.25, 0.46, 0.45, 0.94) 0.33s;
        }
      }
    }
  }
`;
