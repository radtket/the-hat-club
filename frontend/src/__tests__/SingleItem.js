import React from "react";
import { shallow } from "enzyme";
import toJSON from "enzyme-to-json";
import SingleItem from "../components/SingleItem";

const fakeItem = {
  id: "ABC123",
  title: "A Cool Item",
  price: 4000,
  description: "This item is really cool!",
  image: "dog.jpg",
  largeImage: "largedog.jpg",
};

describe("<SingleItem />", () => {
  it("renders and matches the snapshot", () => {
    const wrapper = shallow(<SingleItem {...fakeItem} />);
    expect(toJSON(wrapper)).toMatchSnapshot();
  });

  // it("renders the image properly", () => {
  //   const wrapper = shallow(<SingleItem {...fakeItem} />);
  //   const img = wrapper.find("img");
  //   expect(img.props().src).toBe(fakeItem.image);
  //   expect(img.props().alt).toBe(fakeItem.title);

  //   // console.log(wrapper.debug());
  // });

  // it("renders the pricetag and title", () => {
  //   const wrapper = shallow(<SingleItem {...fakeItem} />);

  //   const PriceTag = wrapper.find("dt");
  //   expect(PriceTag.children().text()).toBe("$40.00");

  //   const Title = wrapper.find("h3");
  //   expect(Title.children().text()).toBe(fakeItem.title);

  //   // console.log(wrapper.debug());
  // });
});
