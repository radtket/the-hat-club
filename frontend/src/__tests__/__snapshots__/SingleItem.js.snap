// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SingleItem /> renders and matches the snapshot 1`] = `
<styled.figure>
  <Link
    to="/item/ABC123"
  >
    <img
      alt="A Cool Item"
      src="dog.jpg"
    />
  </Link>
  <figcaption>
    <h3>
      A Cool Item
    </h3>
    <dl>
      <dt>
        $40.00
      </dt>
      <dd>
        <Link
          to="/item/ABC123"
        >
          View Details
        </Link>
      </dd>
    </dl>
  </figcaption>
</styled.figure>
`;
